### PR TITLE
Feature/233 host error handling

### LINKED ISSUE #233: Graceful Error Handling [3]: Host error handling
Blocked by:  #232 

We need to add some error handling for the Host when things go wrong. Some situations that could happen are:

* Cannot create a lobby
* cannot get cards
* cannot get card-packs
* any edge cases that crash the game

In these cases we need to use the `AlertModal` to let the players and host know something bad happened so they do not see a massive error when something goes wrong. 

image

# User Stories
* As a host, I would like to know if something went wrong in the application that is unrecoverable so I don't have to try to read a massive error message in the dev-tools.

# Completion Criteria
- [x] Places where API requests could fail are covered
- [x] Edge case errors are covered 
   * Look into [error boundaries](https://reactjs.org/docs/error-boundaries.html)
- [x] Clicking "continue" brings the user back to the JOIN/HOST screen.

### PR DESCRIPTION
**Pull request by fortypercenttitanium**
image

#### 0. Make sure your issue is linked:

Close #233 

#### 1. Purpose:

1. Provide an error boundary when a component throws an error during rendering. This shows a user-friendly screen that lets them know something went wrong.
2. Provide custom error messages for all predictable error cases, such as API failures and socket connection issues.

#### 2. Implementation:

* Implements the Alert Modal from #232 
* Sets up an error boundary on the host side similar to #231 
  * This will catch edge-case errors that are difficult to predict right now
* Introduces a new piece of state in the host context for errors. The schema is as follows:
  * error: Object
    * hasError: Boolean (if true, will render the alert modal}
    * errorCallbackType: String (cases will determine what action clicking the modal will perform)
      * 'RESET' - removes the error state and allows the player to continue playing
      * 'RELOAD' (*default*) reloads the page, bringing the user to the home screen.
    * message: Object
      * bigText: String (top text in modal)
      * smallText: String (body text in modal)
      * buttonText: String (button text in modal)
* The alert modal is rendered from the HostLayout component. This preserves the visibility of what was on the user's screen before the error occurred behind the modal background.
* Updates all tests - it was a lot because now the host layout relies on the state error object to know whether or not to render the alert modal

#### 3. Possible Issues:

Explore different edge cases and ways the API can fail. Errors occurring in event listeners will probably remain uncaught, as this is a limitation of the [error boundary in react](https://reactjs.org/docs/error-boundaries.html#introducing-error-boundaries).

### Error boundary
1. Pull branch
2. In `HostScreenController.js` add an error under `case 'waiting-for-lobby':`
```js
case 'waiting-for-lobby':
  throw new Error();
// ...
```
3. Run `npm run dev:frontend`
4. Click "host"
5. Hit ESC (this error won't show in production).
6. Observe the error boundary screen

### Socket error
1. Pull branch
2. Run `npm run dev:frontend`
3. **do not run the backend script**
4. Click "host"
5. Observe the alert modal
6. Alternatively, you can start the game normally and kill the backend once the socket is connected to see the error.

### Fetch error
1. Pull branch
2. In `GameSettings.js`, alter the fetch query URL in the `getPackNames` function (you can misspell "deck")
3. Run `npm run dev:frontend`
4. Run `npm run dev:backend`
5. Click "host"
6. Observe the alert modal

### Fetch error when starting game
1. Pull branch
2. In `hostReducerMiddleware.js`, alter the fetch query URL in the `getDeck` function.
3. Run `npm run dev:frontend`
4. Run `npm run dev:backend`
5. Click "host"
6. Join with 2+ players and click start button
7. Observe the alert modal

### Trying to start game with not enough players or no packs selected
1. Pull branch
2. In `hostReducerMiddleware.js`, alter the fetch query URL in the `getDeck` function.
3. Run `npm run dev:frontend`
4. Run `npm run dev:backend`
5. Click "host"
6. Try to start game with fewer than 2 players, or with no packs selected.

#### 5. Cleanup
- [x] I've added all TODOs needed
- [x] I've removed all unneeded comments
- [x] I've updated the snapshot tests and checked to make sure they're accurate if they changed
- [x] I've removed all unnecessary `console.log`s

### CODE REVIEW

File: frontend/src/errorHandlers/hostErrorHandlers/HostErrorBoundary.spec.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  default: ({ children }) => <div data-testid="modal">{children}</div>,
+}));
+
+describe('error boundary', () => {
+  let spy;
+
+  // we need to mock the console.error or else jest will error out
+  // https://github.com/facebook/react/issues/11098#issuecomment-523977830
+  beforeEach(() => {
+    spy = jest.spyOn(console, 'error').mockImplementation(() => {});
```
[the remark this answers is not in the document]
Comment (fortypercenttitanium) [reply frontend/src/errorHandlers/hostErrorHandlers/HostErrorBoundary.spec.js:23]: I think it's more important to automate the restoration in case any test in the future doesn't need it, and someone forgets to restore it. If this expands (it likely will with error reporting), it's likely every test block should be throwing an error. I'd feel more comfortable leaving it as is.

File: frontend/src/errorHandlers/hostErrorHandlers/HostErrorBoundary.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    super(props);
+    this.state = { hasError: false };
+  }
+
+  static getDerivedStateFromError() {
+    // we use this to set the state from an error message
+    // details can be passed in if we choose in the future
+    // the return value will set the state of this component
+    return {
+      hasError: true,
```
[the remark this answers is not in the document]
Comment (fortypercenttitanium) [reply frontend/src/errorHandlers/hostErrorHandlers/HostErrorBoundary.js:25]: The first test in the spec tests it. The alert modal won't render unless the state flips.

File: frontend/src/screens/HostPregameScreen/HostPregameScreen.js [hunk 3]
```diff
@@ -101,28 +101,70 @@ function LeftPanel() {
 
   const handleClickStart = async () => {
     // check if there are any players and if packs are selected
-    if (playerIDs.length && state.gameSettings.selectedPacks.length) {
+    if (playerIDs.length > 1 && state.gameSettings.selectedPacks.length) {
```
[the remark this answers is not in the document]
Comment (fortypercenttitanium) [reply frontend/src/screens/HostPregameScreen/HostPregameScreen.js:104]: nice idea!

File: frontend/src/contexts/HostContext/hostReducerMiddleware.js [hunk 4]
```diff
@@ -104,13 +104,15 @@ async function getDeck({ selectedPacks }) {
...
+  const errorMessage = `Error fetching cards. Query: ${query}`;
+  const cardsRequest = await fetch(query);
+  if (cardsRequest.ok) {
     const cards = await cardsRequest.json();
-    return cards;
-  } catch {
-    throw new Error(`Error fetching cards. Query: ${query}`);
+    if (cards.white.length) {
+      return cards;
+    }
```
Comment (I3uckwheat) [inline frontend/src/contexts/HostContext/hostReducerMiddleware.js:113]: This is fine for now, but I'm not sure the purpose of this. What if there's no black cards?

**Review by I3uckwheat — CHANGES_REQUESTED** [review]

**Review by wise-king-sullyman — DISMISSED** [review]

File: frontend/src/layouts/HostLayout.spec.js [hunk 5]
```diff
@@ -4,68 +4,95 @@ import userEvent from '@testing-library/user-event';
 import renderer from 'react-test-renderer';
 
 import HostLayout from './HostLayout';
+import { HostContext } from '../contexts/HostContext/HostContext';
```
[the remark this answers is not in the document]
Comment (fortypercenttitanium) [reply frontend/src/layouts/HostLayout.spec.js:7]: Very nice catch, thank you.

**Review by wise-king-sullyman — APPROVED** [review]

**Review by I3uckwheat — APPROVED** [review]
